pkg/metrics/garbagecollector.go

```diff
@@ -0,0 +1,183 @@
+package metrics
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+	nm "github.com/rancher/norman/metrics"
+	"github.com/rancher/rancher/pkg/settings"
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var (
+	targetMetricsByNameForClientKey = []interface{}{
+		nm.TotalAddWS, nm.TotalRemoveWS, nm.TotalAddConnectionsForWS, nm.TotalRemoveConnectionsForWS,
+		nm.TotalTransmitBytesOnWS, nm.TotalTransmitErrorBytesOnWS, nm.TotalReceiveBytesOnWS,
+	}
+
+	targetMetricsByIPForPeer = []interface{}{
+		nm.TotalAddPeerAttempt, nm.TotalPeerConnected, nm.TotalPeerDisConnected,
+	}
+)
+
+type metricGarbageCollector struct {
+	clusterLister  v3.ClusterLister
+	nodeLister     v3.NodeLister
+	endpointLister v1.EndpointsLister
+}
+
+func (gc *metricGarbageCollector) metricGarbageCollection() {
+	logrus.Debugf("[metrics-garbage-collector] Start")
+
+	isClusterMode := settings.Namespace.Get() != "" && settings.PeerServices.Get() != ""
+
+	// Existing resources
+	observedResourceNames := map[string]bool{}
+	// Exisiting Metrics Labels
+	observedLabelsMap := map[string]map[interface{}][]map[string]string{}
+
+	// Get Clusters
+	clusters, err := gc.clusterLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("[metrics-garbage-collector] failed to list clusters: %s", err)
+		return
+	}
+	for _, cluster := range clusters {
+		if _, ok := observedResourceNames[cluster.Name]; !ok {
+			observedResourceNames[cluster.Name] = true
+		}
+	}
+	// Get Nodes
+	nodes, err := gc.nodeLister.List("", labels.Everything())
```

Comment:
think we can call all listers before going over (nodes, clusters)? This way if any of them returns an error, we save cpu on iteration and memory on map allocation :)

Reply:
It only stops if cluster breaks, for the others it keeps going to at least clean clusters. If clusters aren't there there isn't much that can be done.